### ISSUE TITLE
ConcurrentModificationExceptions on config repository updates on GoCD 20.8

### ISSUE THREAD

**Issue by chadlwilson** [gocd/gocd#8711]
##### Issue Type
- Bug Report

##### Summary

Since upgrade from `20.5.0` to `20.8.0` we have started seeing `ConcurrentModificationException`s during periodic config repository updates (without any commit to the config repo). We were not seeing these on `20.5.0`.

They also don't seem to clear from the error message dialog which is frustrating - it's also not clear if the config repo update is entirely stuck/dead. Manually refreshing the config repository doesn't seem to resolve it.

In general there seem to have been config repo regressions somewhere from 20.2.0 to 20.8.0 as we are seeing some problems we never saw on `20.1.0` and there is no clear workaround that we can figure out from the error messages.

###### Basic environment details

* Go Version: `20.8.0`
* JAVA Version: `14.0.2`
* OS: `Linux 4.14.198-152.320.amzn2.x86_64`

###### Additional Environment Details

We have 3 Git config repositories, all being used with the YAML config plugin. There are about 190 pipelines, a number with dependencies between pipelines defined in different repositories that require resolution and merging of the config for pipelines to load. Environments and pipeline groups are also often shared across the 3 config repositories.

One theory is that there are race conditions when all 3 config repos are refreshed at once, even if nothing has changed.

We also have quite a large number of `PluggableScmMaterial`s (~300) added via the [Git Path material plugin](https://github.com/TWChennai/gocd-git-path-material-plugin).

##### Steps to Reproduce

Unclear currently even what the cause is, or whether there is an issue with particular material definitions. It also doesn't seem to be deterministic as to when it occurs, nor how to get it to fix itself.

Any clues on how to narrow down further would be useful.

##### Expected Results
Reliable config repository polls/updates/merging.

##### Actual Results
Getting errors like the below periodically, some time after the server has started; and seemingly after some config repository polls have been successful
```
jvm 1    | java.util.ConcurrentModificationException: null
jvm 1    |      at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(Unknown Source)
jvm 1    |      at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(Unknown Source)
jvm 1    |      at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(Unknown Source)
jvm 1    |      at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
jvm 1    |      at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
jvm 1    |      at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(Unknown Source)
jvm 1    |      at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
jvm 1    |      at java.base/java.util.stream.ReferencePipeline.findFirst(Unknown Source)
jvm 1    |      at com.thoughtworks.go.domain.scm.SCMs.findByFingerprint(SCMs.java:61)
jvm 1    |      at com.thoughtworks.go.domain.scm.SCMs.findDuplicate(SCMs.java:54)
jvm 1    |      at com.thoughtworks.go.config.ConfigConverter.toPluggableScmMaterialConfig(ConfigConverter.java:381)
jvm 1    |      at com.thoughtworks.go.config.ConfigConverter.toMaterialConfig(ConfigConverter.java:319)
jvm 1    |      at com.thoughtworks.go.config.ConfigConverter.toPipelineConfig(ConfigConverter.java:622)
jvm 1    |      at com.thoughtworks.go.config.ConfigConverter.toBasicPipelineConfigs(ConfigConverter.java:116)
jvm 1    |      at com.thoughtworks.go.config.ConfigConverter.toPartialConfig(ConfigConverter.java:85)
jvm 1    |      at com.thoughtworks.go.config.ConfigRepoPlugin.load(ConfigRepoPlugin.java:65)
jvm 1    |      at com.thoughtworks.go.config.GoConfigRepoConfigDataSource.onCheckoutComplete(GoConfigRepoConfigDataSource.ja
va:162)
jvm 1    |      at com.thoughtworks.go.server.materials.ConfigMaterialUpdateListener.updateConfigurationFromCheckout(ConfigMa
terialUpdateListener.java:99)
jvm 1    |      at com.thoughtworks.go.server.materials.ConfigMaterialUpdateListener.onMessage(ConfigMaterialUpdateListener.j
ava:84)
jvm 1    |      at com.thoughtworks.go.server.materials.ConfigMaterialUpdateListener.onMessage(ConfigMaterialUpdateListener.j
ava:36)
jvm 1    |      at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.
java:83)
jvm 1    |      at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java
:63)
jvm 1    |      at java.base/java.lang.Thread.run(Unknown Source)
```

Additional context: Sometimes these eventually seem to lead to spurious errors like the below, which don't make much sense because the error message is complaining about a material duplicating itself. We have checked the material name is not being re-used across multiple config repos or pipelines; nor do there appear to be other materials with the same fingerprint defined elsewhere.

```
1. Cannot save SCM, found multiple SCMs called 'redacted-material'. SCM names are case-insensitive and must be unique.;; 
2. Cannot save SCM, found duplicate SCMs. redacted-material, redacted-material;; 
```

However, we noted that such spurious duplicate SCM/multiple SCM errors were sporadically appearing in `20.5.0` (but **not** earlier on `20.1.0`). On `20.5.0` they seemed to clear themselves eventually, but on `20.8.0` they appear to not be clearing when combined with one of the `ConcurrentModificationException`s above - possibly because there are dependencies between the pipelines in various config repositories?

**Comment by chadlwilson** [comment]
If anyone has a workaround that would avoid a GoCD restart to resolve this temporarily, it would be greatly appreciated.

Once this has happened to one config repo, that repo seems to get stuck (even though the logs, and then the other repos follow it. This leaves us in a broken state. Possibly because of the dependencies between pipelines in config repos?

They also display the below which seems misleading/inaccurate. The config repo was parsed at one point; so not sure why this is displayed:
![image](https://user-images.githubusercontent.com/29788154/97862411-405ed000-1d40-11eb-9036-6df3cf2e511e.png)

Putting into maintenance mode and back again doesn't seem to resolve it.

**Comment by maheshp** [comment]
@chadlwilson we came across the duplicate SCMs issue on build.gocd.org and have pushed a fix https://github.com/gocd/gocd/pull/8653 for the same. 

We will have to take a look at the reasons for the ConcurrentModificationExceptions and try to reproduce it.

**Comment by chadlwilson** [comment]
I wonder if https://github.com/gocd/gocd/commit/9a67ee396becfdfb4e51edef6d17f57a3b45f224#diff-27d82b935e2634f1f5d4e2c8a14a1703024eb8d00f4c3b6ed9dd767a1b3df100R71-R84 is a factor. This would seem to make the ConfigMaterial subsystem to be almost constantly working to re-parse our config repos even when from our perspective there have been no changes?

**Comment by maheshp** [comment]
In GoCD 20.8.0, support for PR/branches was introduced through the [Groovy DSL Plugin](https://github.com/gocd-contrib/gocd-groovy-dsl-config-plugin#support-for-branches-and-prs). The re-parse of the config repo was required in order to support this feature.

I understand this could be expensive. The Config Repo is parsed post the material update. So far, the Config Repo materials did not provide an option to turn off polling. But, with 20.9.0 we are providing an option to [turn off polling](https://github.com/gocd/gocd/pull/8667) and adding support for [webhooks](https://github.com/gocd/gocd/pull/8023) for config repo materials. Do you think this will help in your case?

**Comment by chadlwilson** [comment]
We don't have ability to use webhooks as access to the central source control is uni-directional.

I don't really follow the use case for PRs/branches on the config repo itself (haven't really thought about that) nor why it's necessary to reprocess the whole thing even if nothing has changed. Seems a bit strange, like a build that runs every minute even if you didn't change anything in source control. Seems like a feature that perhaps should be opt-in on config repos, to me?

In any case, any idea why after the ConcurrentModificationException everything might get completely stuck? Any idea of what action might "unstick" it short of a reboot?

This is happening to us roughly every 12 hours which is somewhat fatal, because pipeline changes seem to no longer flow.

**Comment by chadlwilson** [comment]
I wonder if your fix in #8653 may address the same problem here; although unclear to me why we'd only have seen this issue in 20.6 to 20.8. (maybe increased number of checkouts to process increased chance of race conditions + something particular to our setup making reconciling our pipelines/SCMs slower?)

When processing a stream, a `ConcurrentModificationException` implies the sequence underlying the stream has changed underneath it, which looks like it might cause this.

To try and workaround I've set `-Dconfig.material.post.update.threads=1 ` to hopefully remove chance of such issues when processing these. Will see if it helps.

**Comment by maheshp** [comment]
> To try and workaround I've set -Dconfig.material.post.update.threads=1 to hopefully remove chance of such issues when processing these.

Was going to suggest this, guess you understand the internals better. 

> When processing a stream, a ConcurrentModificationException implies the sequence underlying the stream has changed underneath it, which looks like it might cause this.

I had a quick look at the code yesterday, AFAIK the underlying stream should not change since we work with a cached copy of the config. Anyways I will have to take a deeper look to understand why this could happen. I am not sure if #8653 will fix this issue though. I will try to reproduce/understand this issue further and keep you posted.

**Comment by maheshp** [comment]
The changes to parse Config Repo as part of each material update was done as part of GoCD 20.4.0 - https://github.com/gocd/gocd/pull/7949

**Comment by maheshp** [comment]
@chadlwilson I brought up a GoCD server with 300+ SCM materials and multiple ConfigRepos to re-produce this issue.

On **GoCD 20.8.0** I could easily reproduce the `Duplicate SCMs` issue but dint come across the `ConcurrentModificationException`. 

On Experimental build of **GoCD 20.9.0** neither of the issues were reproducible.

I believe the issue with `Duplicate SCMs` existed all along but it shows up now since the ConfigRepos are parsed after each material update. #8653 should resolve this issue.

We are planning a release of GoCD 20.9.0 in the next few days, post that can you upgrade your server and check if it fixes this issue.

**Comment by chadlwilson** [comment]
After 2 days, the `ConcurrentModificationException`s have gone away once configured with only a single `post.update` thread, so definitely seems to be some kind of race condition or shared state between the threads there. I'm not really an expert in the code; but could it be that when partials are merged from one repo while the other repo is doing the duplicate search that it somehow updates shared state in the cached config's `SCMs` list?

My concern was that we were actually seeing the `Duplicate SCMs` issue on earlier builds (20.5.0), but it would clear itself eventually, whereas this one we have only seen on `20.8.0` and it seemed to not be able to fix itself.

**Comment by maheshp** [comment]
@kritika-singh3 and I took a look at the code again today. You are right about the race condition. 
In this case, the `ConcurrentModificationException` is thrown when 2 threads simultaneously update and iterate over an instance variable of the `SCMs` object, check the implementation in 20.8.0 - https://github.com/gocd/gocd/blob/20.8.0/server/src/main/java/com/thoughtworks/go/config/ConfigConverter.java#L381

We think #8653 should fix this issue.

**Comment by chadlwilson** [comment]
Thanks @maheshp and @kritika-singh3 - much appreciated. Fingers crossed 👍

**Comment by chadlwilson** [comment]
Will close this for now and re-open if it doesn't appear to be fixed with `20.9.0`.

**Comment by chadlwilson** [comment]
This does indeed seem to have been fixed on `20.9.0`. Added back the second `config.material.post.update` thread and haven't seen any such errors yet. Will keep an eye on it 👍
